google-cloud-spanner/src/main/java/com/google/cloud/spanner/OpenTelemetryApiTracer.java

```diff
@@ -0,0 +1,276 @@
+      AttributeKey.longKey("attempt.response_count");
+  private final AttributeKey<String> CONNECTION_ID_KEY = AttributeKey.stringKey("connection");
+  private final AttributeKey<Long> RETRY_DELAY_KEY = AttributeKey.longKey("delay_ms");
+  private static final AttributeKey<Long> BATCH_SIZE_KEY = AttributeKey.longKey("batch.size");
+  private static final AttributeKey<Long> BATCH_COUNT_KEY = AttributeKey.longKey("batch.count");
+
+  private final Span span;
+  private final OperationType operationType;
+
+  private volatile String lastConnectionId;
+  private volatile long currentAttemptId;
+  private final AtomicLong attemptSentMessages = new AtomicLong(0);
+  private long attemptReceivedMessages = 0;
+  private final AtomicLong totalSentMessages = new AtomicLong(0);
+  private long totalReceivedMessages = 0;
+
+  OpenTelemetryApiTracer(@Nonnull Span span, @Nonnull OperationType operationType) {
+    this.span = Preconditions.checkNotNull(span);
+    this.operationType = Preconditions.checkNotNull(operationType);
+  }
+
+  Span getSpan() {
+    return this.span;
+  }
+
+  @Override
+  public Scope inScope() {
+    final io.opentelemetry.context.Scope openTelemetryScope = span.makeCurrent();
+    return openTelemetryScope::close;
+  }
+
+  @Override
+  public void operationSucceeded() {
+    span.setAllAttributes(baseOperationAttributes());
+    span.setStatus(StatusCode.OK);
+    span.end();
+  }
+
+  @Override
+  public void operationCancelled() {
+    span.setAllAttributes(baseOperationAttributes());
+    span.setStatus(StatusCode.ERROR, "Cancelled by caller");
+    span.end();
+  }
+
+  @Override
+  public void operationFailed(Throwable error) {
+    span.setAllAttributes(baseOperationAttributes());
+    span.setStatus(StatusCode.ERROR, error.getMessage());
+    span.end();
+  }
+
+  @Override
+  public void lroStartFailed(Throwable error) {
+    span.addEvent(
+        "Operation failed to start", Attributes.of(EXCEPTION_MESSAGE_KEY, error.getMessage()));
+    span.setStatus(StatusCode.ERROR, error.getMessage());
+    span.end();
+  }
+
+  @Override
+  public void lroStartSucceeded() {
+    span.addEvent("Operation started");
+  }
+
+  @Override
+  public void connectionSelected(String id) {
+    lastConnectionId = id;
+  }
+
+  @Override
+  public void attemptStarted(int attemptNumber) {
+    attemptStarted(null, attemptNumber);
+  }
+
+  @Override
+  public void attemptStarted(@Nullable Object request, int attemptNumber) {
+    currentAttemptId = attemptNumber;
+    attemptSentMessages.set(0);
+    attemptReceivedMessages = 0;
+
+    // Attempts start counting a zero, so more than zero indicates a retry.
+    if (attemptNumber > 0 && operationType != OperationType.LongRunning) {
+      // Add an event if the RPC retries, as this is otherwise transparent to the user. Retries
+      // would then show up as higher latency without any logical explanation.
+      span.addEvent("Starting RPC retry " + attemptNumber);
```

Comment:
Have we considered representing retries as child spans? It would definitely make things complicated but maybe easier for customers to understand.

Reply:
I had not considered it before implementing this. But also after having thought about it for a while, I don't think we should do that. The reason is that:
1. We would need to make a choice between 'adding a span for each attempt (including the initial attempt)' and 'adding an extra span only for retry attempts'.
2. The first would mean that we would add an extra span that does not really add any information in the vast majority of the cases when the RPC is not being retried. That would only add additional costs for customers.
3. The second would mean that we would get a varying 'span depth' depending on whether an RPC is being retried or not. I'm not sure that makes it easier to understand for a customer, and I'm worried that it makes it harder for a customer to search for slow requests. With this, a slow request could show up both as a single slow attempt, and as a slow operation with multiple child spans, where the attributes would be spread across the parent and child spans.